src/Farmer/Builders/Builders.DiagnosticSetting.fs

```diff
@@ -0,0 +1,234 @@
+          Retention_period = 0<Days>
+        }
+    member _.Run(state:RetentionPolicy)=
+            match state.Retention_period with
+            | OutOfBounds days -> 
+                failwithf "The retention period must be between 1 and 365 days. It is currently %d" days
+            | InBounds _ -> {state with Enabled = true}
+
+    [<CustomOperation "retention_period">]
+    member _.Days(state:RetentionPolicy, days) = { state with Retention_period=days }
+
+let retentionPolicy=RenentionPolicyBuilder()
+
+type MetricBuilder () = 
+    member _.Yield _ =
+        { Category = ""
+          Enabled = true
+          TimeGrain = None
+          RetentionPolicy= 
+            { Enabled = false
+              Retention_period = 0<Days>  
+            } |> Some
+         }
+
+    /// Sets the Diagnostic Metric category for a resource type.
+    [<CustomOperation "category">]
+    member _.Category(state:MetricSettings, category) = { state with Category=category }
+    
+    /// Sets the timegrain of the metric in ISO8601 format.
+    [<CustomOperation "time_grain">]
+    member _.TimeGrain(state:MetricSettings, timeGrain) = { state with TimeGrain=Some timeGrain }
+
+    ///  The retention in days metric setting. Must be between 1 and 365 days. 0 is selected by default.                         
+    [<CustomOperation "retention_period">]
+    member _.Retention_period(state: MetricSettings, retentionPeriod: int<Days>) = 
+        { state with 
+            RetentionPolicy =Some (retentionPolicy { retention_period retentionPeriod }) }
+
+type LogBuilder () = 
+    member _.Yield _ =
+        { Category = ""
+          Enabled = true
+          RetentionPolicy=
+            { Enabled = false
+              Retention_period = 0<Days>  
+            } |> Some 
+         }
+
+    /// Sets the Diagnostic Log category for a resource type 
+    [<CustomOperation "category">]
+    member _.Category(state: LogSettings, category) = { state with Category = category }
+
+     ///  The retention in days for log settings. Must be between 1 and 365 days. 0 is selected by default.                      
+    [<CustomOperation "retention_period">]
+    member _.Retention_period(state: LogSettings, retentionPeriod: int<Days>) = 
+        {state with 
+            RetentionPolicy = Some (retentionPolicy { retention_period retentionPeriod }) }
+
+type DiagnosticSettingsBuilder() =
+    member _.Yield _ =
+         { Name = ResourceName.Empty
+           StorageAccountId = None
+           ServiceBusRuleId = None
+           EventHubAuthorizationRuleId = None
+           EventHubName = None
+           Metrics = []
+           Logs = []
+           Dependencies=[]
+           ParentResourceType = ""
+           WorkspaceId = None
+           DedicatedLogAnalyticsDestination = None
+           Tags = Map.empty }
+    member _.Run(state:diagnosticSettingsConfig) =
+            
+        match state.EventHubName,state.EventHubAuthorizationRuleId with 
```

Comment:
You could do single set of pattern matches over state using Record matching e.g.

```fsharp
match state with
| { EventHubName = Some _; EventHubAuthorizationRuleId = None } -> failwith....
| { Metrics = []; Logs = [] } -> failwith ....
...
| _ -> state

```

Reply:
Wow this is a nice way to do it